### PR TITLE
DataGrid: Rework the headerFilter.dataSource article

### PR DESCRIPTION
**Pull request by RomanTsukanov**
Fixes the following docissues:
- https://supportcenter.devexpress.com/internal/ticket/details/T878805
- https://supportcenter.devexpress.com/internal/ticket/details/T879477
- https://supportcenter.devexpress.com/internal/ticket/details/T874405
- https://supportcenter.devexpress.com/internal/ticket/details/T866708
- https://supportcenter.devexpress.com/internal/ticket/details/T855896

### CODE REVIEW

**Review by artem-kurchenko — APPROVED** [review]

**Comment by artem-kurchenko** [comment]
@RomanTsukanov  a side question/suggestion 
Should we mention Doc Issues ids in the Description, so we can understand what topics are going to be covered?

**Comment by RomanTsukanov** [comment]
> 
> 
> @RomanTsukanov a side question/suggestion
> Should we mention Doc Issues ids in the Description, so we can understand what topics are going to be covered?

Done

**Review by onebalaban — APPROVED** [review]
OK

**Review by AlexJameson — APPROVED** [review]

**Review by vconst — APPROVED** [review]

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ default: undefined
 ---
 ---
 ##### shortDescription
-Specifies a data source for the header filter.
+Specifies the header filter data source.
```
Comment (dirk-pieterse) [inline api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]:
```suggestion
Specifies the header filter's data source.
```

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md [hunk 2]
```diff
@@ -14,242 +14,207 @@ Data source options.
 The widget's instance.
 
 ##### field(options.dataSource): DataSource_Options
-The header filter's data source.
+A [DataSource](/Documentation/ApiReference/Data_Layer/DataSource/) instance.
 
 ---
-This option accepts one of the following values:
+The **{WidgetName}** generates a header filter data source automatically based on column values. Use the **dataSource** option to change the generated data source or specify a custom data source.
```
Comment (dirk-pieterse) [inline api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]:
```suggestion
The **{WidgetName}** generates a header filter's data source automatically based on column values. Use the **dataSource** option to change the generated data source or specify a custom data source.
```

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md [hunk 3]
```diff
@@ -14,242 +14,207 @@ Data source options.
...
 
-- **Array of Objects**      
-A simple JavaScript array containing a collection of plain objects. 
+### Specify a Custom Data Source
 
-    ---
-    #####jQuery
+To define a data source from scratch, set the **dataSource** option to an array of objects. Each object configures one header filter item and should have the following fields:
```
Comment (dirk-pieterse) [inline api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]: "To define a data source from scratch" Can't you just replace it with 'to define a data source'? 

```suggestion
To define a data source, set the **dataSource** option to an array of objects. Each object configures one header filter item and should have the following fields:
```

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md [hunk 4]
```diff
@@ -14,242 +14,207 @@ Data source options.
...
 
-        <!--HTML--><dx-{widget-name} ...>
-            <dxi-column>
-                <dxo-header-filter [dataSource]="headerFilterData"></dxo-header-filter>
-            </dxi-column>
-        </dx-{widget-name}>
+### Map Data Source Fields
 
-    ---
+Header filter data objects should have the `text` and `value` fields, but when the data objects are fetched from a server, you have no control over their fields. In this case, map the original data source to the `text` + `value` structure. A mapped data source should also include key fields from the original data source.
```
Comment (dirk-pieterse) [inline api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]:
```suggestion
Header filter data objects should have the `text` and `value` fields. However, data objects fetched from a server have fields assigned to them automatically. In this case, map the original data source to the `text` + `value` structure. A mapped data source should also include key fields from the original data source.
```

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md [hunk 5]
```diff
@@ -309,18 +286,518 @@ If your data source omits the **text** and **value** fields, you can cast it to
...
+
+---
+
+#include common-demobutton with {
+    url: "https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/Filtering/"
+}
+
+### Customize Item Appearance
+
+You can use [templates](/Documentation/ApiReference/Common/Object_Structures/template/) to customize the appearance of individual items. In the following code, a template applies the bold font to the `Today` item:
```
Comment (dirk-pieterse) [inline api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]:
```suggestion
You can use [templates](/Documentation/ApiReference/Common/Object_Structures/template/) to customize the appearance of individual items. In the following code, a template makes the `Today` item bold:
```

**Review by dirk-pieterse — APPROVED** [review]

File: api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md (hunk 4 above)
Comment (RomanTsukanov) [reply api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]:
```suggestion
Header filter data objects should have the `text` and `value` fields. However, data objects fetched from a server may not have these fields. In this case, map the original data source to the `text` + `value` structure. A mapped data source should also include key fields from the original data source.
```

Comment (dirk-pieterse) [reply api-reference/_hidden/GridBaseColumn/headerFilter/dataSource.md]: Ok

**Pull request merged**